compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/serviceprovider/GlobalAtomicLong.java

```diff
@@ -61,8 +79,31 @@ public class GlobalAtomicLong {
      *
      * @param initialValue initial value to which the long is set when its memory is allocated
      */
-    public GlobalAtomicLong(long initialValue) {
+    public GlobalAtomicLong(String name, long initialValue) {
+        this.name = name;
         this.initialValue = initialValue;
+        if (ImageInfo.inImageRuntimeCode()) {
+            throw GraalError.shouldNotReachHere("Cannot create " + getClass().getName() + " objects in native image runtime");
+        } else {
+            if (ImageInfo.inImageBuildtimeCode() && LibGraalSupport.INSTANCE != null) {
+                addressSupplier = LibGraalSupport.INSTANCE.createGlobal(initialValue);
+            } else {
+                addressSupplier = () -> {
+                    if (ImageInfo.inImageRuntimeCode()) {
+                        throw GraalError.shouldNotReachHere("The addressSupplier field value should have been replaced at image build time");
+                    }
+                    long addr = UNSAFE.allocateMemory(Long.BYTES);
```

Comment:
Add a comment that this code-path is used when running as graal.jar (not as native-image)

Reply:
Done.